### ISSUE TITLE
Router client disconnect() before recv() causing crash (assertion failed)

### ISSUE THREAD

**Issue by cenzhao** [zeromq/libzmq#3485]
*Please use this template for reporting suspected bugs or requests for help.*

# Issue description
TLDR;
sequence to reproduce the issue
router(client) <-> router(server) communication.
- server send 3 frames to client: <client-identity-string> <empty-string> <dummy-content>
- client side: poll until data ready to be read.
- client make a zmq_disconnect()
- client do zmq_recv() 3 times, first 2 times got correct data, 3rd call -> "Assertion failed: !_more (src/fq.cpp:112) Aborted (core dumped)"

Detailed Use Cases:
I have an event loop that handles two events:
(1): data ready to read on socket-1
(2): received a command on socket-2 to disconnect a peer. 
event loop keeps looping by calling zmq_poll() over these 2 sockets and handle them accordingly. 

if two events happened at the same time, the handling order is undetermined. If I happen to receive zmq-message form peer-X AND got a command to disconnect() peer-X. 

this following would crash the program:
zmq_poll() indicate two events ready.
zmq_disconnect(peer-X) (process event(2) first)
zmq_recvmsg() (process event(1)) -> assertion failed -> program crash.

based on documentation: http://api.zeromq.org/4-3:zmq-disconnect , if disconnect() called before poll(), poll() will block since message got discarded. 
But if poll() is called BEFORE disconnect(), we should still be able to read the data, shoudn't crash the program.

# Environment

* libzmq version: 4.3.1
* OS: ubuntu 16.04

# Minimal test code / Steps to reproduce the issue

    // create ROUTER client and server socket
    void *server = test_context_socket (ZMQ_ROUTER);
    void *client = test_context_socket (ZMQ_ROUTER);

    // set identity
    zmq_setsockopt(server, ZMQ_IDENTITY, "server", 6);
    zmq_setsockopt(client, ZMQ_IDENTITY, "client", 6);

    // server bind, client connect, let server send 3 frames: <client-id><><dummy-content>
    zmq_bind(server, "inproc://server");
    zmq_connect(client, "inproc://server");
    char msg1[] = "client";
    char msg2[] = "";
    char msg3[] = "test"; 
    zmq_send(server, msg1, 6, ZMQ_SNDMORE);
    zmq_send(server, msg2, 1, ZMQ_SNDMORE);
    zmq_send(server, msg3, 5, 0);

    // poll on client until data ready to read
    std::cout << "poll ..." << std::endl;
    zmq_pollitem_t items[1];
    items[0].socket = client; 
    items[0].events = ZMQ_POLLIN;
    zmq_poll (items, 1, -1); 
    std::cout << "client data ready to read ..." << std::endl;

    // client disconnect to server
    zmq_disconnect(client, "inproc://server");

    // read 3 frames out
    char buf[256];
    int len1 = zmq_recv(client, buf, 256, ZMQ_DONTWAIT); 
    std::cout << "len1: " << len1 << ":" << buf << std::endl;
    int len2 = zmq_recv(client, buf, 256, ZMQ_DONTWAIT); 
    std::cout << "len2: " << len2 << ":" << buf << std::endl;
    int len3 = zmq_recv(client, buf, 256, ZMQ_DONTWAIT); 
    std::cout << "len3: " << len3 << ":" << buf << std::endl;

# What's the actual result? (include assertion message & call stack if applicable)

poll ...
client data ready to read ...
len1: 6:server
len2: 1:
Assertion failed: !_more (src/fq.cpp:112)
Aborted (core dumped)

first 2 reads got data, 3rd read -> assertion failed. program crashed.
# What's the expected result?
we should still be able to read the 3-framed message without crashing the program.

**Comment by sigiesec** [comment]
The documentation of zmq_disconnect says nothing about zmq_poll, only about zmq_recv:
```
Any outstanding messages physically received from the network but not
yet received by the application with _zmq_recv()_ shall be discarded.
```
Since the disconnect is asynchronous, this cannot refer to the return from the function, but to the point of time when the disconnect is performed in the I/O thread. The documentation should probably be amended to reflect this.

So the pending message might or might not be still receivable via zmq_recv after the call to zmq_disconnect returned, depending on timing.

It is not completely clear, if a multi-part message might be possibly received only partially, but since a multi-part message is consider atomic in other contexts, the atomicity should probably not be violated here.

Another question is if after zmq_poll with ZMQ_POLLIN indicated that a message is ready to be received, it is required that this message can still be received via zmq_recv, as you suggest. I am not sure if this can be implemented.

In any case, no crash/assertion failure should result, so there is definitely a bug.

@bluca Have you got an idea how this might be solved? Where are the messages associated with a specific endpoint cleared?

**Comment by saifhhasan** [comment]
Any update guys. cc @bluca .. We're hitting this issue intermittently in production environment at Facebook.
